server/api_conversation.go

```diff
@@ -0,0 +1,118 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/pluginapi"
+
+	"github.com/gin-gonic/gin"
+)
+
+func (p *Plugin) handlePostConversation(c *gin.Context) {
+	userID := c.GetHeader("Mattermost-User-Id")
+
+	// We only allow bots to use this API handler for the time being.
+	if _, err := p.pluginAPI.Bot.Get(userID, false); errors.Is(err, pluginapi.ErrNotFound) {
+		c.AbortWithError(http.StatusForbidden, errors.New("forbidden"))
+		return
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get bot: %w", err))
+		return
+	}
+
+	var conv []*model.Post
+	if err := json.NewDecoder(c.Request.Body).Decode(&conv); err != nil {
+		c.AbortWithError(http.StatusBadRequest, err)
+		return
+	}
+	defer c.Request.Body.Close()
+	if len(conv) == 0 {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty conversation"))
+		return
+	}
+
+	// The last post is expected to contain the current request, everything else is used as context.
+	post := conv[len(conv)-1]
+
+	if post.ChannelId == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty channel id"))
+		return
+	}
+
+	channel, err := p.pluginAPI.Channel.Get(post.ChannelId)
+	if errors.Is(err, pluginapi.ErrNotFound) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("channel not found"))
+		return
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get channel: %w", err))
+		return
+	}
+
+	if post.UserId == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty user id"))
+		return
+	}
+
+	postingUser, err := p.pluginAPI.User.Get(post.UserId)
+	if errors.Is(err, pluginapi.ErrNotFound) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("user not found"))
+		return
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get posting user: %w", err))
+		return
+	}
+
+	// Don't respond to ourselves
+	if p.IsAnyBot(post.UserId) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("not responding to ourselves"))
+		return
+	}
+
+	bot := p.GetBotMentioned(post.Message)
```

Comment:
I am keeping the message requirements intact, such as needing to mention a specific bot.

Reply:
Maybe we should take the bot name as a parameter instead of requiring a mention in the post?

Reply:
Yeah, given https://github.com/mattermost/mattermost-plugin-ai/pull/188#discussion_r1619525511, I think it makes sense to make this a bit more structured. It may also be cleaner to decouple the current request/prompt from any past context.